mcs/class/corlib/Test/System.IO/FileTest.cs

```diff
@@ -2680,5 +2683,73 @@ public void MoveTest ()
 			MoveTest (FileAccess.ReadWrite, FileShare.Write | FileShare.Delete, true);
 			MoveTest (FileAccess.ReadWrite, FileShare.ReadWrite | FileShare.Delete, true);
 		}
+
+
+		[DllImport ("libc", SetLastError=true)]
+		public static extern int symlink (string oldpath, string newpath);
+
+		[DllImport ("libc", SetLastError=true)]
+		public static extern int rename (string oldpath, string newpath);
+
+		/* This deserves some examplation:
+		Calling File.Delete on a bad symlink will throw UnauthorizedAccessException so they are highlander files to managed code.
+		The hack here is to break the cycle and make them dangling symlinks instead.
+		 */
+		static void DeleteBadSymLink (string path) {
+			string newpath;
+
+			lock (random) {
+				do {
+					newpath = path + random.Next ();
+				} while (File.Exists (newpath));
+			}
+
+			rename (path, newpath);
+			File.Delete (newpath);
+		}
+
+		[Test]
+		public void SymLinkLoop ()
+		{
+			if (!RunningOnUnix)
+				Assert.Ignore ("Symlink are hard on windows");
+			string name1, name2;
+			lock (random) {
+				name1 = "first_file" + random.Next ();
+				name2 = "second_file" + random.Next ();
+			}
+
+			var path1 = Path.Combine (Path.GetTempPath (), name1);
+			var path2 = Path.Combine (Path.GetTempPath (), name2);
+
+			DeleteBadSymLink (path1);
+			DeleteBadSymLink (path2);
+			try {
+				symlink (path1, path2);
+				symlink (path2, path1);
+
+				Assert.IsFalse (File.Exists (path1), "File.Exists must return false for symlink loops");
+
+				try {
+					using (var f = File.Open (path1, FileMode.Open, FileAccess.Read)) {
+						Assert.Fail ("File.Exists must throw for symlink loops");
+					}
+				} catch (UnauthorizedAccessException ex) {
+					Assert.AreEqual (0x80070005u, (uint)ex.HResult, "Ensure HRESULT is correct");
+				}
+
+
+				try {
+					File.Delete (path1);
```

Comment:
this should have Assert.Fail() otherwise we won't notice if this stops throwing.